Collisions

erickurquilla opened the pull request:
All tests on Jenkins were passed. Now the code evolves N_ab instead of N and f_ab. Some analysis scripts might still need changes.

Commit:
Particles store the phase space volume they occupy

Commit:
set evolution equations for N and L to be zero in code generation script in anticipation of changing them based on opacities. Set Vphase to not evolve in time.

Commit:
whitespace

Commit:
infrastructure to set IMFPs and chemical potentials in the parameter file

Commit:
redefining f --> Nrho (before was f ---> Nrho)

Commit:
computing volume of phase space

Commit:
deleting lines that convert twice mu to erg

Commit:
adding delta E as an input parameter to compute phase space volume

Commit:
adding new input parameters to input files to run collitions

Commit:
adding collition term in QKE

Commit:
moving line that compute derivates, now it set the phase volume to zero meaning its time derivative its zero after it is used in collision term

Commit:
considering the particle that carry antineutrinos in the fast flavor test

Commit:
Fixing an error in the collision term

Commit:
including adaptative time step for absortion and emission

Commit:
creating vacuum paticles initial condition script

Commit:
adding attenuation parameter to the time derivative of N_Ab due to the hamiltonians

Commit:
solving mistake to compute time step due to collisions

Commit:
adding collision test input file

Commit:
testing if the particles converge to the expected equilibrium distribution

Commit:
solving issue to copy GNUmakefile_jenkins to Exec directory

Commit:
doing fast compilation in collision test

Commit:
clarifying better the collision time steps, now delta t collision is collision_CFL / ( c * max_IMFP )

Commit:
deletting renomarlization function. It will not be used anymore

Commit:
solving mistake in compilation comand to compile the code for 3 flavors

srichers commented on Scripts/symbolic_hermitians/generate_code.py:
```diff
@@ -461,6 +452,38 @@ def sgn(t,var):
+        if(t == ""): s=0
+        if(t == "bar"): s=1
+
+        code.append(["dfdt00_Re"+t+" += IMFP_abs["+str(s)+"][0] * f_eq["+str(s)+"][0] * p.rdata(PIdx::Vphase) * pow( 1 / ( 2 * MathConst::pi * PhysConst::hbar ) , 3 ) * ( 1 / PhysConst::c2 ) - PhysConst::c * ( ( IMFP_abs["+str(s)+"][0] * f_eq["+str(s)+"][0] + IMFP_abs["+str(s)+"][0] * f_eq["+str(s)+"][0] ) / 2 + ( IMFP_abs["+str(s)+"][0] + IMFP_abs["+str(s)+"][0] ) / 2 ) * p.rdata(PIdx::f00_Re"+t ...
```
We should assume the IMFP already includes final-state blocking, so those terms should not be explicitly stated here.

srichers commented on Scripts/symbolic_hermitians/generate_code.py:
```diff
@@ -469,9 +492,17 @@ def sgn(t,var):
+        line = "p.rdata(PIdx::N"+t+") = 0;"
+        code.append([line])
+
+        # evolution equations for L and Lbar, stored as dLdt-->L
```
Is the evolution equation for L really zero?

srichers commented on Scripts/symbolic_hermitians/generate_code.py:
```diff
@@ -469,9 +492,17 @@ def sgn(t,var):
         # Write out dFdt->F
         code.append(dFdt.code())
 
+        # evolution equations for N and Nbar, stored as dNdt-->N
```
Is the evolution equation for N zero? Should we even be storing N at this point?

srichers commented on Source/Evolve.cpp:
```diff
@@ -62,13 +62,32 @@ Real compute_dt(const Geometry& geom, const Real cfl_factor, const MultiFab& sta
-	Real dt_flavor_stupid   = PhysConst::hbar/Vmax_stupid  *flavor_cfl_factor;
+	Real dt_flavor_adaptive   = PhysConst::hbar/Vmax_adaptive*flavor_cfl_factor;
+	Real dt_flavor_stupid     = PhysConst::hbar/Vmax_stupid  *flavor_cfl_factor;
+	Real dt_flavor_absorption = PhysConst::hbar/Vmax_stupid  *flavor_cfl_factor; // it will change only if opacity_method is 1
```
This should be set to something like infinity to keep it from playing a role unless activated below. Giving it some reasonable value could allow it to accidentally be used in the wrong way.

srichers commented on Source/Evolve.cpp:
```diff
@@ -62,13 +62,32 @@ Real compute_dt(const Geometry& geom, const Real cfl_factor, const MultiFab& sta
+        // Calculate dt_flavor_absorption
+        dt_flavor_absorption = (1 / (PhysConst::c * max_IMFP_abs)) * parms->collision_cfl_factor;
+
+        if (parms->attenuation_hamiltonians==0){
```
I think this could be resolved by increasing dt_flavor (both values) by a factor of 1/attenuation_hamiltonian. That would make the timestep correctly scale with the attenuation parameter, too.

srichers commented on Source/Evolve.cpp:
```diff
@@ -62,13 +62,32 @@ Real compute_dt(const Geometry& geom, const Real cfl_factor, const MultiFab& sta
+    dt_flavor = dt_flavor_stupid;
 	if(max_adaptive_speedup>1)
-	  dt_flavor = min(dt_flavor_stupid*max_adaptive_speedup, dt_flavor_adaptive);
+	  dt_flavor = min(dt_flavor_stupid*max_adaptive_speedup, dt_flavor_adaptive, dt_flavor_absorption);
```
The timestep should be limited by absorption even if the max adaptive speedup is larger than 1. I think this min should be present outside of the if statement.

srichers commented on Source/Parameters.H:
```diff
@@ -42,7 +42,17 @@ struct TestParams : public amrex::Gpu::Managed
+    int IMFP_method;
+    Real IMFP_abs[2][NUM_FLAVORS];
+    Real IMFP_scat[2][NUM_FLAVORS];
+    Real munu[2][NUM_FLAVORS]; // equilibrium electron neutrino chemical potential in MeV
```
This should be in CGS units, right?

srichers commented on Source/Parameters.H:
```diff
@@ -42,7 +42,17 @@ struct TestParams : public amrex::Gpu::Managed
+    Real IMFP_abs[2][NUM_FLAVORS];
+    Real IMFP_scat[2][NUM_FLAVORS];
+    Real munu[2][NUM_FLAVORS]; // equilibrium electron neutrino chemical potential in MeV
+    Real delta_E;
```
Comment the units.

srichers commented on Source/main.cpp:
```diff
@@ -197,7 +194,7 @@ void evolve_flavor(const TestParams* parms)
         // because the last deposit_to_mesh call was at either the old time (forward Euler)
         // or the final RK stage, if using Runge-Kutta.
-        const Real dt = compute_dt(geom,parms->cfl_factor,state,neutrinos,parms->flavor_cfl_factor,parms->max_adaptive_speedup);
+        const Real dt = compute_dt(geom,parms->cfl_factor,state,neutrinos,parms->flavor_cfl_factor,parms->max_adaptive_speedup, parms);
```
I would prefer to just pass parms as a whole and delete the individual parms arguments

srichers commented on Source/main.cpp:
```diff
@@ -206,7 +203,7 @@ void evolve_flavor(const TestParams* parms)
 
     // Get a starting timestep
-    const Real starting_dt = compute_dt(geom,parms->cfl_factor,state,neutrinos_old,parms->flavor_cfl_factor, parms->max_adaptive_speedup);
+    const Real starting_dt = compute_dt(geom,parms->cfl_factor, state,neutrinos_old,parms->flavor_cfl_factor, parms->max_adaptive_speedup, parms);
```
Same here - I would just delete the individual parms arguments.

Commit:
Setting default collision time step to infinite. It will change if the code do collisions.

Commit:
setting right indentation

Commit:
including attenuation parameter to compute the time steps due to potentials

Commit:
computing minimum between time steps due to potentials and collision even if max_adaptive_speedup>1 is not satisfy

Commit:
solve mistake in chemical potential units in coment

Commit:
add coment on the energy units

Commit:
delete the individual parms arguments

Commit:
creating collisional flavor instability test

Commit:
delete print line

Commit:
now reads plt* files for time steps of more than 5 digits

Commit:
now reads plt files for time steps with more than 5 digits

Commit:
add function to initialize the background enviroment Ye, rho and T

Commit:
insert chemical potential for neutrinos and antineutrino independently

Commit:
set python script that generate initial condition for collisional instability test

Commit:
set input file for collisional instability test

Commit:
Create a Jenkins test for collisional flavor instability

Commit:
change cuda version to 7.0

Commit:
delete std cout line

Commit:
correct cuda version for jenkins

Commit:
change order of commands first reduce_data_fft.py then reduce_data.py

Commit:
delete function to initialize background quantities on this branch

Commit:
Merge branch 'collisions' of https://github.com/AMReX-Astro/Emu into collisions

Commit:
Solve mistake on collisional instability test. Now it read the right input file

Commit:
Solve mistake on collisional instability test. Now it read the right python script

Commit:
add python script for collision instability test

Commit:
solve problem of defining chemical potentials for antineutrinos in the input parameter file

Commit:
Merge remote-tracking branch 'origin/development' into collisions

Commit:
rename f -> N for particles

Beware: grid also has a Nij variable which is different

Commit:
finish renaming f -> N everywhere

Commit:
comment lines that assert if N and Nbar are negatives, this lines will be deleted since now N and Nbar will no longer be used

Commit:
change f for N in utility script to generate initial particles conditions

Commit:
modify initial condition script for msw test. now it gives N_ab instead of N and f_ab as input

Commit:
add comments to the function that read the initial conditions of the particles

Commit:
change msw test to include the change f_ab ---> N_ab

Commit:
reading N_ab for the particle initial condition file instead of N and f_ab independenly

Commit:
removing f_ab variables for N_ab when setting perturbation

Commit:
delete no needed lines

Commit:
repare coll_equi_test.py test to consider N_ab instead of f_ab

Commit:
making bypolar test work with the changes f_ab ---> N_ab

Commit:
making fast flavor test work for f_ab ---> N_ab

Commit:
making fast_flavor_k_test works for f_ab ---> N_ab

Commit:
delete print line

Commit:
delete print lines

Commit:
make attenuation parameter symbolic in code generation scripts

srichers commented on Scripts/symbolic_hermitians/generate_code.py:
```diff
@@ -445,86 +434,62 @@ def sgn(t,var):
         Gdeclare = ["amrex::Real {}".format(line) for line in G.code()]
         code.append(Gdeclare)
 
+        #collision term (emmission and apsortion)
```
I think this should be written symbolically to take advantage of the power of sympy for more maintanable and extendable code

Commit:
move f-->N in a few more places

Commit:
change f to N in test scripts

Commit:
change Trf to TrN

Commit:
remove unused line

srichers commented on Source/Parameters.H:
```diff
@@ -74,28 +88,58 @@ struct TestParams : public amrex::Gpu::Managed
+    	pp.get("delta_E", delta_E);
+        delta_E*= 1e6*CGSUnitsConst::eV; // erg
+    }
+    else AMREX_ASSERT_WITH_MESSAGE(false, "only available opacity_method is 1");
```
This should be 0 or 1, right?

srichers reviewed (changes requested):
The only changes are a small overlooked comment and making the code generation script actually generate the code symbolically without if/else statements

erickurquilla committed:
setting right initial conditions for collisional instability test

Commit:
inteporlate T, Ye, rho from grid to particles, but return right value multiplied by 4

Commit:
solve mistake in interpolating rho, Ye and T from grid to particles posision, now the generation code script does not repeate the code 4 times

Commit:
delete print lines

Commit:
using the interpolated value of T for the calculation of the equlibrium distribution

Commit:
collisional instability test do not include a random perturbation on the off diagonal component of the density matrix. Not it use vacuum potentials.

erickurquilla committed:
Merge branch 'collisions' of https://github.com/AMReX-Astro/Emu into collisions

erickurquilla committed:
upload julien nsm input script

erickurquilla committed:
solve mistake in energy average

Commit:
set right initial conditions for julien nsm simulation

Commit:
Modify collision instability test

The collision instaiblity test now includes the LSA in equation 14 of L. Johns [2104.11369]
Julien Frousley script to solve isotropic QKE is used to compare EMU solution
Plots are included all over the test but were commented

Commit:
Add collisions directory to store usefull script for simulations containing collisions

The script compute_dE.py is used to compute the energy bin size base on the expected equilibrium number of neutrinos

Commit:
Comments added to compute_dE.py script

Commit:
reordered scripts in jenkinsfile to clean up PR

Commit:
added line in data reduction scripts to prevent glob from picking up generated hdf5 files

srichers commented on Scripts/collisions/compute_dE.py:
```diff
@@ -0,0 +1,42 @@
+'''
+This script compute the energy bin size ( /Delta E ) for monoenergetic neutrino simulations, given:
```
write explicitly which test this is used for

srichers commented on Scripts/initial_conditions/st9_jul_nsm.py:
```diff
@@ -0,0 +1,36 @@
+import h5py
```
This would be good to keep in the directory running the simulations - more code in the repository is a larger obligation to maintain, so less code is better if possible.

Commit:
Deleting unnecessary variables in the script that generates initial conditions for the collision to equilibrium test (coll_equi_test.py).

This commit reduces code size by removing redundant variables in the st7_empty_particles.py script (Scripts/tests/coll_equi_test.py), which generates a set of particles with zero neutrinos and antineutrinos.

srichers commented on Scripts/symbolic_hermitians/generate_code.py:
```diff
@@ -437,6 +426,12 @@ def sgn(t,var):
             code.append(line)
             code.append("")
+        
+    code.append("T_pp += sx(i) * sy(j) * sz(k) * sarr(i, j, k, GIdx::T);")
```
rho and ye should be interpolated only once. Currently rho*ye is interpolated to the particle position. Instead, we should use these values of rho and ye, and only later construct the potential from the interpolated values. This ensures that all thermodynamic quantities are consistent between collisions and flavor transformation.

srichers replied:
On second thought, I think we should move ahead with this PR and come back if necessary.

srichers commented on Scripts/tests/coll_inst_test.py:
```diff
@@ -0,0 +1,253 @@
+    ######################################################################################
+    ######################################################################################
+
+    """
```
This should be in a separate script, not saved in the repo unless continuously useful.

srichers commented on Scripts/tests/convergence.sh:
```diff
@@ -57,7 +57,7 @@ do_convergence () {
 INTPARAMS="
 integration.type=0"
 
-INTNAME="fe"
```
This is an old script from Don when we were testing convergence. This file can be deleted.

srichers commented on Scripts/tests/plot_convergence.py:
```diff
@@ -69,12 +69,12 @@ def plot_on_axis(self, axis, key, label, color):
             
```
remove this file as well.

srichers commented on Source/Evolve.cpp:
```diff
@@ -146,6 +164,11 @@ void interpolate_rhs_from_mesh(FlavoredNeutrinoContainer& neutrinos_rhs, const M
         const ParticleInterpolator<SHAPE_FACTOR_ORDER> sz(delta_z, shape_factor_order_z);
 
+        // The following variables contains temperature, electron fraction, and density interpolated from grid quantities to particle positions
+        double T_pp = 0;
```
use Real instead of double

srichers commented on Source/Parameters.H:
```diff
@@ -42,7 +42,17 @@ struct TestParams : public amrex::Gpu::Managed
+
+    // absorption opacities and equilibrium neutrino chemical potentials
+    int IMFP_method;
+    Real IMFP_abs[2][NUM_FLAVORS];
```
remove these variables. When using IMFP_method==0, just access the params struct.

srichers commented on sample_inputs/inputs_1d_fiducial:
```diff
@@ -78,3 +82,27 @@ alpha2_degrees = 0
+#################
+IMFP_method = 0
+
+IMFP_abs0_cm = 0
```
remove IMFP variables if IMFP_method==0

srichers commented on sample_inputs/inputs_bipolar_test:
```diff
@@ -78,3 +82,28 @@ alpha2_degrees = 0
+#################
+IMFP_method = 0
+
+IMFP_abs0_cm = 0
```
remove IMFP variables if IMFP_method==0

srichers commented on sample_inputs/inputs_fast_flavor:
```diff
@@ -78,3 +82,28 @@ alpha2_degrees = 0
+#################
+IMFP_method = 0
+
+IMFP_abs0_cm = 0
```
remove IMFP variables if IMFP_method==0

srichers commented on sample_inputs/inputs_fast_flavor_nonzerok:
```diff
@@ -79,3 +83,27 @@ alpha2_degrees = 0
+#################
+IMFP_method = 0
+
+IMFP_abs0_cm = 0
```
remove IMFP variables if IMFP_method==0

srichers commented on sample_inputs/inputs_jul_nsm:
```diff
@@ -0,0 +1,111 @@
+perturbation_type = 0
```
file should be deleted unless continuously useful. Keep the initial conditions and parameter file in the directory where you do the simuation

srichers commented on sample_inputs/inputs_msw_test:
```diff
@@ -78,3 +82,28 @@ alpha2_degrees = 0
+#################
+# opacity stuff #
+#################
+IMFP_method = 0
```
remove IMFP variables if IMFP_method==0

srichers commented on Scripts/symbolic_hermitians/generate_code.py:
```diff
@@ -445,86 +440,54 @@ def sgn(t,var):
+        #collision term (emmission and apsortion)
+        s=0
+        if(t == ""): s=0
+        if(t == "bar"): s=1
```
Add option to include "stimulated absorption" in reaction rates

Commit:
Delete Julien's NSM input file from the initial conditions script directory.

Commit:
Removed opacities and chemical potential entries from input files that do not include collision simulations.

To avoid future confusion, I deleted the collision-related input options (opacities, chemical potential, and energy bin size) from input files used for tests without collisions.

Commit:
Added a more accurate description of the assert error for opacity methods.

The assert error now clearly indicates that the only available opacity methods are 0 (no collisions) and 1 (with collisions).

Commit:
Change variable type from double to Real to keep consistency

Commit:
Adding more comments to the script that computes the energy bin size for collision tests.

Commit:
Delete commented lines that are under development.

I deleted commented lines containing Julien's script that computes the collisional flavor instability evolution. This could potentially be included in the future to solidify the collisional instability test.

Commit:
Delete Python script of a previous test that will no longer be used.

Commit:
Delete Python script that generates the initial conditions of particles in Julien's NSM simulation.

Commit:
Delete old script used for testing convergence. It will no longer be used.

Commit:
Change N_eq to f_eq since this better represents the collision term.

f_eq is the Fermi-Dirac neutrino distribution for a given particle energy.

Commit:
Include option to multiply the inverse mean free paths by the Pauli blocking term.

A new input parameter was created (Do_Pauli_blocking). If set to 1, it will multiply the inverse mean free path by 1 / (1 - f_eq); if set to 0, it will do nothing.

Commit:
Adding a semicolon

Commit:
Adding the input parameter Do_Pauli_blocking to all input files.

I added the input parameter Do_Pauli_blocking to all input files to avoid errors caused by not reading this parameter during execution. If this parameter is set to 1, it will multiply the inverse mean free path by 1 / (1 - f_eq); if set to 0, it will do nothing. In the tests that don't involve collisions, it will not have any effect.

Commit:
Correction in the collision term of QKE

The collision term implemented in this commit is C = { gamma , N_eq - N }. Here Gamma = diag( k*_e , k*_x ) / 2 . k*_a is the inverse mean free path for flavor a, including Pauli blocking term. * means that Pauli blocking term is already in the inverse mean free path values. Equilibrium neutrino number matrix N_eq equals the integral of f_eq, where the integral is over the phase space that the particle represents.

Commit:
Modify the script to compute the energy bin size and antineutrino chemical potential for the collision instability test.

This change emerged from the correction made to the collision term and the necessity of obtaining the antineutrino chemical potential to ensure the test runs properly.

Commit:
Setting up the collisional instability test

Set the appropriate input parameters for the collisional instability test, generate plots to visualize the results, and add comments to the Python script that generates the initial conditions for this test.

srichers commented:
Is it possible to speed up the collisions test? It is currently taking 25 minutes on Jenkins.

srichers committed:
Merge branch 'development' into collisions
